agent/eventhandler/attachment_handler.go

```diff
@@ -0,0 +1,158 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package eventhandler
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/aws/amazon-ecs-agent/agent/api"
+	"github.com/aws/amazon-ecs-agent/agent/statechange"
+	"github.com/aws/amazon-ecs-agent/agent/statemanager"
+	"github.com/aws/amazon-ecs-agent/agent/utils/retry"
+	"github.com/cihub/seelog"
+)
+
+// AttachmentEventHandler is a handler that is responsible for submitting attachment state change events
+// to backend
+type AttachmentEventHandler struct {
+	// stateSaver is a statemanager which may be used to save any
+	// changes to an attachment's SentStatus
+	stateSaver statemanager.Saver
+
+	// backoff is the backoff object used in submitting attachment state change
+	backoff retry.Backoff
+
+	// attachmentLock is a map from attachment ARN to the attachmentHandler that is
+	// responsible for handling the attachment
+	attachmentARNToHandler map[string]*attachmentHandler
+
+	// lock is used to safely access the attachmentARNToHandler map
+	lock sync.Mutex
+
+	client api.ECSClient
+	ctx    context.Context
+}
+
+// attachmentHandler is responsible for handling a certain attachment
+type attachmentHandler struct {
+	// attachmentARN is the arn of the attachment that the attachmentHandler is handling
+	attachmentARN string
+
+	// stateSaver is a statemanager which may be used to save any
+	// changes to an attachment's SentStatus
+	stateSaver statemanager.Saver
+
+	// backoff is the backoff object used in submitting attachment state change
+	backoff retry.Backoff
+
+	// lock is used to ensure that the attached status of an attachment won't be sent multiple times
+	lock sync.Mutex
+
+	client api.ECSClient
+	ctx    context.Context
+}
+
+// NewAttachmentEventHandler returns a new AttachmentEventHandler object
+func NewAttachmentEventHandler(ctx context.Context,
+	stateSaver statemanager.Saver,
+	client api.ECSClient) *AttachmentEventHandler {
+	return &AttachmentEventHandler{
+		ctx:                    ctx,
+		stateSaver:             stateSaver,
+		client:                 client,
+		attachmentARNToHandler: make(map[string]*attachmentHandler),
+		backoff: retry.NewExponentialBackoff(submitStateBackoffMin, submitStateBackoffMax,
+			submitStateBackoffJitterMultiple, submitStateBackoffMultiple),
+	}
+}
+
+// AddStateChangeEvent adds a state change event to AttachmentEventHandler for it to handle
+func (handler *AttachmentEventHandler) AddStateChangeEvent(change statechange.Event) error {
```

Comment:
can you rename this to be `eventHandler`? Makes it less confusing to read the code.

Reply:
updated